### PR TITLE
fix(types): make definitions `nodenext` compatible

### LINKED ISSUE #223: Incorrect default import types in ESM
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.6.0

### Plugin version

8.1.0

### Node.js version

18.7.0

### Operating system

Windows

### Operating system version (i.e. 20.04, 11.3, 10)

10

### Description

When importing `@fastify/cors` as ESM, TypeScript has the incorrect type definions for the default import, forcing the user to add `.default` to the import, even though just using the default import (`cors` in the case) works at runtime

image

image

I believe this is a typing issue, since the runtime works perfectly in each case

### Steps to Reproduce

- Have a TypeScript project with the following configuration:
    ```json
    {
      "compilerOptions": {
        "target": "es2022",
        "lib": ["es2022"],

        "module": "esnext",
        "moduleResolution": "nodenext",
        "baseUrl": "src",
        "resolveJsonModule": true,

        "sourceMap": true,
        "outDir": "dist",
        "pretty": true,
        "noEmitOnError": true,

        "allowSyntheticDefaultImports": true,
        "esModuleInterop": true,
        "forceConsistentCasingInFileNames": true,

        "strict": true,

        "skipLibCheck": true
      }
    }
  ```
- Import `@fastify/cors`
- Try to register it normally
- Run the TypeScript compiler

### Expected Behavior

Assuming the plugin was imported as `cors`, using `fastify.register(cors, ...)` should compile successfully, it throws an error instead

### PR DESCRIPTION
**Pull request by SuperchupuDev**
Fixes #223

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Uzlopak** [comment]
Do we have a typing test for this?

**Comment by Uzlopak** [comment]
CI is failing.

**Review by climba03003 — CHANGES_REQUESTED** [review]
It is not a proper fix and it will just break the existing user.
The proper fix should follow https://github.com/fastify/fastify-cookie/pull/184 or https://github.com/fastify/fastify-static/pull/311

And mostly it requires to fix in all modules.

**Comment by SuperchupuDev** [comment]
I already tried fixing the tests but the error messages are really complex for a person that doesn't know much about fastify's internals.

File: index.d.ts [hunk 1]
```diff
@@ -3,76 +3,88 @@
...
+     * Hide options route from the documentation built using fastify-swagger (default: true).
+     */
+    hideOptionsRoute?: boolean;
+  }
+
+  export interface FastifyCorsOptionsDelegateCallback { (req: FastifyRequest, cb: (error: Error | null, corsOptions?: FastifyCorsOptions) => void): void }
+  export interface FastifyCorsOptionsDelegatePromise { (req: FastifyRequest):  Promise<FastifyCorsOptions> }
+  export type FastifyCorsOptionsDelegate = FastifyCorsOptionsDelegateCallback | FastifyCorsOptionsDelegatePromise
+  export type FastifyPluginOptionsDelegate<T = FastifyCorsOptionsDelegate> = (instance: FastifyInstance) => T;
+
```
Comment (climba03003) [inline index.d.ts:82]:
```suggestion

export { fastifyCors }
```

**Review by mcollina — APPROVED** [review]
lgtm

@fox1t can you confirm this is correct?

File: index.d.ts [hunk 2]
```diff
@@ -3,76 +3,89 @@
...
+export const fastifyCors: FastifyCorsPlugin
+  export { fastifyCors as default };
```
Comment (fox1t) [inline index.d.ts:84]: Please add also the named export since we export the famous triplet in JS.
`export { fastifyCors };`

**Review by fox1t — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 2 above)
Comment (climba03003) [reply index.d.ts:84]: Line 83 is named export

**Review by climba03003 — APPROVED** [review]
LGTM after Lint fix

File: index.d.ts (hunk 2 above)
Comment (fox1t) [reply index.d.ts:84]: Oh my bad! It is a lint issue!

**Review by fox1t — APPROVED** [review]

File: index.d.ts [hunk 3]
```diff
@@ -3,76 +3,89 @@
...
+     */
+    hideOptionsRoute?: boolean;
+  }
+
+  export interface FastifyCorsOptionsDelegateCallback { (req: FastifyRequest, cb: (error: Error | null, corsOptions?: FastifyCorsOptions) => void): void }
+  export interface FastifyCorsOptionsDelegatePromise { (req: FastifyRequest):  Promise<FastifyCorsOptions> }
+  export type FastifyCorsOptionsDelegate = FastifyCorsOptionsDelegateCallback | FastifyCorsOptionsDelegatePromise
+  export type FastifyPluginOptionsDelegate<T = FastifyCorsOptionsDelegate> = (instance: FastifyInstance) => T;
+
+export const fastifyCors: FastifyCorsPlugin
```
Comment (Uzlopak) [inline index.d.ts:83]:
```suggestion
  export const fastifyCors: FastifyCorsPlugin
```

**Pull request merged**